File: adapters/rxjava2/src/main/java/io/fotoapparat/result/adapter/rxjava2/SingleAdapter.java
package io.fotoapparat.result.adapter.rxjava2;

import java.util.concurrent.Future;

import io.fotoapparat.result.adapter.Adapter;
import io.reactivex.Single;

/**
 * Adapter for {@link Single}.
 */
public class SingleAdapter<T> implements Adapter<T, Single<T>> {

    private SingleAdapter() {
    }

    /**
     * @return {@link Adapter} which adapts result to {@link Single}.
     */
    public static <R> SingleAdapter<R> toSingle() {
        return new SingleAdapter<>();
    }

    @Override
    public Single<T> adapt(Future<T> future) {
        return Single.fromFuture(future);
    }
}

File: fotoapparat-adapters/rxjava2/src/main/java/io/fotoapparat/result/adapter/rxjava2/SingleAdapter.kt
package io.fotoapparat.result.adapter.rxjava2

import android.annotation.SuppressLint
import io.fotoapparat.result.PendingResult
import io.reactivex.Single
import java.util.concurrent.Future

/**
 * Adapter for [Single].
 */
object SingleAdapter {

    /**
     * @return Adapter which adapts result to [Single].
     */
    @JvmStatic
    @SuppressLint("CheckResult")
    fun <T> toSingle(): (Future<T>) -> Single<T> {
        return { future -> Single.fromFuture(future) }
    }

}

/**
 * @return A [Single] from the given [PendingResult].
 */
@SuppressLint("CheckResult")
fun <T> PendingResult<T>.toSingle(): Single<T> {
    return adapt { future -> Single.fromFuture(future) }
}
